addon/components/listbox.js

```diff
@@ -0,0 +1,298 @@
+import Component from '@glimmer/component';
+import { tracked } from '@glimmer/tracking';
+import { guidFor } from '@ember/object/internals';
+import { action } from '@ember/object';
+import { debounce } from '@ember/runloop';
+
+const ACTIVATE_NONE = 0;
+const ACTIVATE_FIRST = 1;
+const ACTIVATE_LAST = 2;
+
+export default class ListboxComponent extends Component {
+  @tracked activeOptionIndex;
+  activateBehaviour = ACTIVATE_NONE;
+  buttonElement;
+  guid = `${guidFor(this)}-headlessui-listbox`;
+  @tracked _isOpen = this.args.isOpen || false;
+  labelElement;
+  optionsElement;
+  optionElements = [];
+  optionValues = {};
+  search = '';
+  @tracked selectedOptionIndex;
+
+  get activeOptionGuid() {
+    return this.optionElements[this.activeOptionIndex]?.id;
+  }
+
+  get isDisabled() {
+    return !!this.args.disabled;
+  }
+
+  get selectedOptionGuid() {
+    return this.optionElements[this.selectedOptionIndex]?.id;
+  }
+
+  get isOpen() {
+    return this._isOpen;
+  }
+
+  set isOpen(isOpen) {
```

Comment:
I really like using a "setter" like this; keeps the code for manipulating the value clean, but allows us to perform these side-effects easily as well!